pkg/providers/vsphere/virtualmachine/cleanup.go

```diff
@@ -0,0 +1,164 @@
+// © Broadcom. All Rights Reserved.
+// The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: Apache-2.0
+
+package virtualmachine
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
+	vimtypes "github.com/vmware/govmomi/vim25/types"
+
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+)
+
+// CleanupVMServiceState removes all modifications made by VM Operator
+// from a vCenter VM. This should be called when deleting a VM
+// resource with the skip-delete-platform-resource annotation to
+// ensure the vCenter VM is left in a clean state.
+//
+// For now, this method only cleans up the ExtraConfig keys that we
+// add and the managedBy property. We should enhance this to clean any
+// tags, cluster module associations etc that were added as part of
+// VM's lifecycle. But those are risky and require some discussion, so
+// handle that in a followup.
+func CleanupVMServiceState(
+	vmCtx pkgctx.VirtualMachineContext,
+	vimClient *vim25.Client,
+	vcVM *object.VirtualMachine) error {
+
+	vmCtx.Logger.Info("Sanitizing vCenter VM")
+
+	var moVM mo.VirtualMachine
+	if err := vcVM.Properties(
+		vmCtx,
+		vcVM.Reference(),
+		[]string{"config"},
+		&moVM); err != nil {
+
+		return fmt.Errorf("failed to fetch vm properties for cleanup: %w", err)
+	}
+
+	if moVM.Config == nil {
+		vmCtx.Logger.Info("VM config is nil, skipping cleanup")
+		return nil
+	}
+
+	configSpec := vimtypes.VirtualMachineConfigSpec{}
+
+	// Remove ExtraConfig keys
+	removeVMOperatorExtraConfig(moVM.Config, &configSpec)
+
+	// Clear ManagedBy field
+	clearManagedBy(moVM.Config, &configSpec)
+
+	// Apply the configuration changes
+	if err := reconfigureVM(vmCtx, vcVM, configSpec); err != nil {
+		return fmt.Errorf("failed to apply cleanup configuration: %w", err)
+	}
+
+	vmCtx.Logger.Info("Successfully sanitized the vCenter VM")
+	return nil
+}
+
+// removeVMOperatorExtraConfig identifies and marks for removal all
+// ExtraConfig keys that were added by VM Operator.
+//
+// Unlike FilteredExtraConfig which preserves allow-listed keys for reuse
+// scenarios, this method removes ALL VM Operator keys (both deny-listed
+// and allow-listed) since we're unregistering the VM from Supervisor.
+func removeVMOperatorExtraConfig(
+	config *vimtypes.VirtualMachineConfigInfo,
+	configSpec *vimtypes.VirtualMachineConfigSpec) {
+
+	currentEC := object.OptionValueList(config.ExtraConfig)
+
+	// Get the deny prefixes and allow keys from extraconfig.go
+	// We need to remove keys matching BOTH deny prefixes AND allow list
+	denyPrefixes := []string{
+		"guestinfo.",
```

Comment:
You cannot assume we should delete *all* of the ones that match guest info. That's too broad.

Reply:
So, the starting point for this was the extraConfig sanitization that we do where we trim all fields that have a [prefix](https://github.com/vmware-tanzu/vm-operator/blob/a5e6ef2ddd04269a49724c339a3b1b0a5e65c026/pkg/providers/vsphere/virtualmachine/extraconfig.go#L24-L28).